### ISSUE TITLE
GeoTiff: public hasAlphaBand() instead of addAlpha_?

### ISSUE THREAD

**Issue by m-mohr** [openlayers/openlayers#13643]
**Is your feature request related to a problem? Please describe.**
I'd like to know whether an alpha band has been added to the GeoTiff bands or not. I see that there is a property `addAlpha_`, but this sounds like a private property and not a public API. 

**Describe the solution you'd like**
I'm wondering whether it would be good to add a public API method such as `hasAlphaBand()` that simply returns `this.addAlpha_`? Happy to create a PR if that's they way to go, but also open to other solutions. Maybe I'm missing something, too.

**Comment by ahocevar** [comment]
An alpha band will only be added by OpenLayers when you have `nodata` configured for the source. There is no magic here, so probably no need to add such a method.

**Comment by m-mohr** [comment]
@ahocevar I don't think I agree here. I don't have nodata configured (myself in code), so the nodata must be coming from the source file. As such, I don't know whether an alpha band has been added or not, but I need to know. So I either need to know whether the alpha band has been added or whether a nodata value has been set. Looking at the code, I also don't see a public access method for the no data values (i.e. `nodataValues_` is private). So what is the proposed way to get the no data value (or whether the alpha band has been added?)

**Comment by ahocevar** [comment]
@m-mohr Sorry for the premature reply, you're absolutely right that the `nodata` value can also be set from the `geotiff.js` image's `getGDALNoData()` method.

I agree it does make sense to be able to access the nodata values. But for use in an application, it is not only enough to know the nodata value, but also the band index that holds it. So to come up with a meaningful API here, it would be good to know a bit more about your use case. Do you need the nodata value for the layer style, or for something completely different?

**Comment by m-mohr** [comment]
I'm basically writing a general-purpose GeoTiff Viewer, i.e. you pass a URL and then the Viewer tries it best to visualize it. For this, it offers the user a list of available bands (either from STAC metadata or if not present there, from the source). For this I'm reading the bandCount property, but sometimes I need to reduce it by one as there is an transparency layer added. So my code looks something like:

```js
			if (this.bands.length === 0) {
				this.bands = Utils.range(1, geotiff.bandCount).map(id => ({ id, min: null, max: null }));
				// Remove the alpha band if present at the end
				if (this.data.addAlpha_) {
					this.bands.pop();
				}
			}
```

I mean it works, but I'd prefer to not access private properties in code. :-)

Similarly, I'm accessing also some other properties and that's why I also +1'ed on the getMetadata issue: #13612

**Comment by ahocevar** [comment]
It looks like with your use case, you'd need to make more or different decisions than OpenLayers makes in the GeoTIFF source. So if you are reading STAC metadata, wouldn't it make sense for you to also read metadata and other information from the GeoTIFF directly using `geotiff.js`? This is also what I suggested in #13612. That way, you could also handle the case where a different band than the last one is the alpha band.

**Comment by m-mohr** [comment]
@ahocevar Yes, mabe. To be more concrete, this is what I'm looking for right now. I'd like to know the actual number of bands and per band:
- whether it is "data" or "alpha"
- min and max values (the ones actually used by OL)
- no data values (the ones actually used by OL)
- band name, if present

That helps to generate a UI that includes defaults, which users can tweak then.

Actually, it seems neither `getMatadata()` in #13625 nor your proposal using geotiff.js really cater for this completely as it just returns the plain metadata from the file. It doesn't take into account when min/max values are determined by OL based on the data type, for example. I assume my use case would need something like an event-based system that emits once the data is available?

As a fallback, I used STAC metadata, but that's also not always present.

**Comment by ahocevar** [comment]
So, this is a sensible requirement, but I don't think it gets you all the way there. However, I think it could be addressed with a data object that explains the bands of the data (i.e. output) tiles:
```js
[{
  sourceIndex: 0,
  sourceBand: 0,
  bandName: 'red',
  min: 0,
  max: 255
}, {
  sourceIndex: 0,
  sourceBand: 1,
  bandName: 'green',
  min: 0,
  max: 255
}, {
  sourceIndex: 0,
  sourceBand: 2,
  bandName: 'blue',
  min: 0,
  max: 255
}, {
  sourceIndex: 0,
  sourceBand: 3,
  alpha: true
}]
```
The problem is that most of this information can be specified using the [SourceInfo](https://openlayers.org/en/latest/apidoc/module-ol_source_GeoTIFF.html#~SourceInfo) configuration object, and OpenLayers only takes information from the GeoTIFF source if these are not provided. So in the end, I think you need additional metadata from each GeoTIFF already by the time you construct the `GeoTIFF` source, not afterwards. Because you may end up needing to configure the GeoTIFF source differently.

**Comment by m-mohr** [comment]
@ahocevar Yes, such a data object would help a lot.

I don't specify any information in the source info right now (except `normalize: false` and the URL of the source). What I do then is to apply styling once I know the min/max values, basically doing linear stretching. no-data I also tried to do via styling, but that did not work yet. But I run into some issues with that approach. For some file it works, for some not (yet?).

I'm also stumbling across some parts of the documentation, where I'm not sure they are always correct. For example, in https://openlayers.org/en/latest/apidoc/module-ol_style_expressions.html#~ExpressionValue it says the value you get for bands is 0..1, but is that also the case if normalize is false? And what exactly does convertToRGB do... it makes it work for grayscale, but then fails for me when I'm loading RGB files, which is weird because I'd assume converting from RGB to RGB would basically do nothing. :-) But I'm slowly digging deeper...

**Comment by duncan1987** [comment]
@m-mohr ，I'm totally agree with you about the style expressions debuging,   any suggestions now?
